nexus/db-queries/src/db/pool.rs

```diff
@@ -3,108 +3,153 @@
+        bootstrap_dns,
+        DnsResolverConfig {
+            hardcoded_ttl: Some(tokio::time::Duration::MAX),
+            ..Default::default()
+        },
+    ))
 }
 
-#[derive(Clone, Debug)]
-struct LoggingErrorSink {
-    log: slog::Logger,
+fn make_single_host_resolver(
+    config: &DbConfig,
+) -> qorb::resolver::BoxedResolver {
+    Box::new(SingleHostResolver::new(config))
 }
 
-impl LoggingErrorSink {
-    fn new(log: slog::Logger) -> LoggingErrorSink {
-        LoggingErrorSink { log }
-    }
+fn make_postgres_connector(
+    log: &Logger,
+) -> qorb::backend::SharedConnector<QorbConnection> {
+    // Create postgres connections.
+    //
+    // We're currently relying on the DieselPgConnector doing the following:
+    // - Disallowing full table scans in its implementation of "on_acquire"
+    // - Creating async_bb8_diesel connections that also wrap DTraceConnections.
+    let user = "root";
+    let db = "omicron";
+    let args = Some("sslmode=disable");
+    Arc::new(DieselPgConnector::new(
+        log,
+        DieselPgConnectorArgs { user, db, args },
+    ))
 }
 
-impl bb8::ErrorSink<ConnectionError> for LoggingErrorSink {
-    fn sink(&self, error: ConnectionError) {
-        error!(
-            &self.log,
-            "database connection error";
-            "error_message" => #%error
-        );
+impl Pool {
+    /// Creates a new qorb-backed connection pool to the database.
+    ///
+    /// Creating this pool does not necessarily wait for connections to become
+    /// available, as backends may shift over time.
+    pub fn new(log: &Logger, bootstrap_dns: Vec<SocketAddr>) -> Self {
+        // Make sure diesel-dtrace's USDT probes are enabled.
+        usdt::register_probes().expect("Failed to register USDT DTrace probes");
+
+        let resolver = make_dns_resolver(bootstrap_dns);
+        let connector = make_postgres_connector(log);
+
+        let policy = Policy::default();
+        Pool { inner: qorb::pool::Pool::new(resolver, connector, policy) }
+    }
+
+    /// Creates a new qorb-backed connection pool to a single instance of the
+    /// database.
+    ///
+    /// This is intended for tests that want to skip DNS resolution, relying
+    /// on a single instance of the database.
+    ///
+    /// In production, [Self::new] should be preferred.
+    pub fn new_single_host(log: &Logger, db_config: &DbConfig) -> Self {
+        // Make sure diesel-dtrace's USDT probes are enabled.
+        usdt::register_probes().expect("Failed to register USDT DTrace probes");
+
+        let resolver = make_single_host_resolver(db_config);
+        let connector = make_postgres_connector(log);
+
+        let policy = Policy::default();
+        Pool { inner: qorb::pool::Pool::new(resolver, connector, policy) }
+    }
+
+    /// Creates a new qorb-backed connection pool which returns an error
+    /// if claims are not quickly available.
+    ///
+    /// This is intended for test-only usage.
```

Comment:
it might be nice if this comment explained when test code should use `new_single_host` and when it should use this constructor?

Reply:
Sure thing, added.